### PR TITLE
Broaden casting behaviour in `data_color()` to support classed colour vectors

### PR DESCRIPTION
**Pull request by mccarthy-m-g**
# Summary

This PR broadens casting behaviour in `data_color()` to support classed colour vectors as inputs to `palette`. The purpose of this change is to make **gt** friendlier to work with when using colour palettes with a custom class, such as those made with the [palettes](https://mccarthy-m-g.github.io/palettes/) package (disclaimer: I'm the author).

Currently, classed colour vectors must first be cast to character vectors with `as.character()`in order to work with `data_color()`; they cannot be supplied directly because the underlying code from the **scales** package that gt relies on doesn't have methods for custom colour classes (including those from the paletteer package). This PR fixes this by casting all inputs to `palette` to a character vector, no longer treating the paletteer palettes as an edge case.

The changes are minimal:

- One `as.character()` call was moved to a new location in `data_color()`
- The `palette` argument's documentation was updated to reflect that it now supports classed colour vectors that can be cast to a character vector as input
- Tests were added for classed colour vectors

I wasn't sure if this change was worth listing in NEWS. Please let me know if you'd like me to add it in.

# Related GitHub Issues and PRs

- Ref: #1155

# Checklist

- [x] I understand and agree to the [Code of Conduct](https://www.contributor-covenant.org/version/2/1/code_of_conduct.html).
- [ ] I have listed any major changes in the [NEWS](https://github.com/rstudio/gt/blob/master/NEWS.md).
- [x] I have added [`testthat`](https://github.com/r-lib/testthat) unit tests to [`tests/testthat`](https://github.com/rstudio/gt/tree/master/tests/testthat) for any new functionality.

### CODE REVIEW

**Comment by olivroy** [comment]
Hi @mccarthy-m-g. If you are still interested, would you mind resolving merge conflicts, sign the CLA and maybe provide a visual example of the feature you want to add here.

## Behavior before

```r
# working code (or workaround needed)
```

< screenshot of result (if applicable)

## This PR

```r
# working code
```

< insert screenshot of result>

Thanks!

**Comment by mccarthy-m-g** [comment]
Hi @olivroy, I've resolved the merge conflicts and resigned the CLA.

Here's some pseudo-code giving a visual example of the feature added in this PR.

## Behaviour before

Right now users have to apply a workaround to use a classed colour vector with the `palette` argument of `data_color()` by casting it to a character vector. You can see a working example of this in the [compatibility with gt vignette](https://mccarthy-m-g.github.io/palettes/articles/gt.html#adding-colour) I wrote for the palettes package.

```r
my_colours <- palettes::pal_colour(c("red", "blue"))

gt_tbl %>% 
  data_color(
    columns = col,
    palette = as.character(my_colours)
  )
```

< A gt table with a coloured column >

## This PR

Users can now pass the classed colour vector directly to the `palette` argument. No workarounds needed.

```r
my_colours <- palettes::pal_colour(c("red", "blue"))

gt_tbl %>% 
  data_color(
    columns = col,
    palette = my_colours
  )
```

< A gt table with a coloured column >

File: R/data_color.R [hunk 1]
```diff
@@ -835,9 +836,16 @@ data_color <- function(
...
-      palette <- as.character(paletteer::paletteer_d(palette = palette))
+      # `paletteer::paletteer_d()`
+      palette <- paletteer::paletteer_d(palette = palette)
     }
+
+    # Cast the palette to a character vector for compatibility with
+    # scales:::toPaletteFunc(), which does not have methods for classed color
+    # vectors (such as those in the paletteer package). Casting is done here so
+    # classed color vectors from packages other than paletteer can also be used
+    # with `palette`.
```
Comment (olivroy) [inline R/data_color.R:847]:
```suggestion
    # with `palette` (#1155). 
```

File: gt.Rproj [hunk 2]
```diff
@@ -21,6 +21,4 @@ PackageUseDevtools: Yes
 PackageInstallArgs: --no-multiarch --with-keep.source
 PackageRoxygenize: rd,collate,namespace
 
-UseNativePipeOperator: No
-
 SpellingDictionary: en_US
```
Comment (olivroy) [inline gt.Rproj:24]:
```suggestion
UseNativePipeOperator: No

SpellingDictionary: en_US
```

Reverting the change

File: tests/testthat/test-data_color.R [hunk 3]
```diff
@@ -358,6 +358,76 @@ test_that("The correct color values are obtained when defining a palette", {
     gsub("(.*: |;$)", "", .) %>%
     expect_in(c("#000000", "#FFFFFF"))
 
+  # Create a `rgba_hex_colors_mixed_classed` object by using `data_color` with
```
Comment (olivroy) [inline tests/testthat/test-data_color.R:361]:
```suggestion
})

test_that("data_color() works with classed colors (#1155)", {
  # Create a `rgba_hex_colors_mixed_classed` object by using `data_color` with
```

File: tests/testthat/test-data_color.R [hunk 4]
```diff
@@ -358,6 +358,76 @@ test_that("The correct color values are obtained when defining a palette", {
...
+    render_as_html() %>%
+    xml2::read_html()
+
+  # Expect 12 unique color values to have been generated and used
+  tbl_html_hex_colors_mixed_classed %>%
+    selection_value("style") %>%
+    gsub("(background-color: |; color: .*)", "", .) %>%
+    unique() %>%
+    length() %>%
+    expect_equal(12)
```
Comment (olivroy) [inline tests/testthat/test-data_color.R:385]:
```suggestion
    expect_length(12)
```

File: tests/testthat/test-data_color.R [hunk 5]
```diff
@@ -358,6 +358,76 @@ test_that("The correct color values are obtained when defining a palette", {
...
+  # since the `autocolor_text` option is TRUE (the default case)
+  (
+    (tbl_html_hex_colors_mixed_classed %>%
+       selection_value("style") %>%
+       gsub("(.*: |;$)", "", .)) %in%
+      c("#000000", "#FFFFFF")
+  ) %>%
+    all() %>%
+    expect_true()
+
```
Comment (olivroy) [inline tests/testthat/test-data_color.R:430]:
```suggestion
})

test_that("Palettes work with data_color()", {
```

**Review by olivroy — APPROVED** [review]
Thanks for the explanations. This seems reasonable. I have left a few comments in the tests. @rich-iannone, once this is resolved, you can run workflows and merge

**Comment by mccarthy-m-g** [comment]
Thanks! I applied, then subsequently revised the suggested changes. The original suggestions created some unintended failures in the tests due to splitting up the test environments, so I put the tests that should have remained in the same environment back together.

**Comment by olivroy** [comment]
Thanks for the oversight!

**Review by rich-iannone — APPROVED** [review]
LGTM!

**Comment by rich-iannone** [comment]
Thank you @mccarthy-m-g for the contribution and really sorry for the long wait in getting this one merged!

**Comment by mccarthy-m-g** [comment]
No worries, thanks for merging!

**Pull request merged**